### PR TITLE
Allow auto-converting of attribute values with a `convert` arg to attr.ib()

### PR DESCRIPTION
**Pull request by radix**
Fixes: #26 

Pretty straightforward change. I decided to go with running convert post-validator. Bikeshedding welcome.

### CODE REVIEW

**Comment by hynek** [comment]
On an Italian beach so only briefly:

- change log please
- two spaces after periods
- an example pls

Thanks! :)

**Comment by radix** [comment]
I think I may have convinced myself that conversion should go before validation, not after.

For example, say I have a validator that wants to make sure that it receives a `PMap` with `"name"` and `"id"` keys. I would want it to happen after conversion, so that my plain dict could be converted to a `PMap` before validation.

However, this would mean that converters may end up raising exceptions for bad inputs instead of validators. For example, if you pass an integer instead of a mapping, the converter will just blow up when trying to convert it to a pmap, instead of the validator (perhaps) raising a nicer exception. I think this is probably acceptable, because I think it's more important to be able to get data into the right shape before validation happens.

I'm gonna change that in my branch.

File: docs/changelog.rst [hunk 1]
```diff
@@ -16,7 +16,7 @@ The third digit is only for regressions.
 Changes:
 ^^^^^^^^
 
-None
+- Add a `convert` argument to :func:`attr.ib`, which allows specifying a function to run on arguments.  This allows for simple type conversions, e.g. with ``attr.ib(convert=int)``. `[26] <https://github.com/hynek/attrs/issues/26>`
```
Comment (hynek) [inline docs/changelog.rst]: semantic new lines pls

File: docs/examples.rst [hunk 2]
```diff
@@ -329,6 +329,43 @@ You can also disable them globally:
    TypeError: ("'x' must provide <InterfaceClass __builtin__.IFoo> which 42 doesn't.", Attribute(name='x', default=NOTHING, validator=<provides validator for interface <InterfaceClass __builtin__.IFoo>>, repr=True, cmp=True, hash=True, init=True), <InterfaceClass __builtin__.IFoo>, 42)
 
 
+Conversion
+----------
+
+Attributes can have a ``convert`` function specified, which will be called with
```
Comment (hynek) [inline docs/examples.rst]: semantic new lines pls

File: docs/examples.rst [hunk 3]
```diff
@@ -329,6 +329,43 @@ You can also disable them globally:
...
+.. doctest::
+
+    >>> @attr.s
+    ... class C(object):
+    ...     x = attr.ib(convert=int)
+    >>> o = C("1")
+    >>> o.x
+    1
+
+Converters are run _before_ validators, so you can use validators to check the
```
Comment (hynek) [inline docs/examples.rst]: semantic new lines pls

File: docs/examples.rst [hunk 4]
```diff
@@ -329,6 +329,43 @@ You can also disable them globally:
...
+final form of the value.
```
Comment (hynek) [inline docs/examples.rst]: semantic new lines pls

File: attr/_make.py [hunk 5]
```diff
@@ -81,6 +82,12 @@ def attr(default=NOTHING, validator=None,
 
     :param init: Include this attribute in the generated ``__init__`` method.
     :type init: bool
+
+    :param callable convert: :func:`callable` that is called by
```
Comment (hynek) [inline attr/_make.py]: aw I forgot about this syntax! could you convert the remaining arguments to it too for consistency please?

**Comment by hynek** [comment]
I wonder, wouldn’t it be better to make the signature of converts the same as for validators? (i.e. passing the attribute and the instance along with the value)  It gives more flexibility and it’s nice to have some symmetry.

**Comment by hynek** [comment]
*bump*

I would like to do a release with #29, it would be nice if we could get this in too?

**Comment by radix** [comment]
@hynek I think the signature should be just the value to maximize the ability to use existing functions of the form `f(a) -> b` in the Python ecosystem (e.g. built-in functions like `int`, `frozenset`, and things like `pyrsistent.freeze`). I could imagine perhaps adding some introspection code to allow three-argument functions but I don't really see a use case for them now.

Sorry for dropping this for so long... I'll update the branch

**Pull request closed without merging**